File: server/converters/stream.ts
/**
 * Gemini ストリームイベント → Claude SSE イベント変換
 *
 * Gemini SDK の sendStream() が生成する ServerGeminiStreamEvent を、
 * Claude Messages API の SSE (Server-Sent Events) 形式に変換する。
 */

import { randomUUID } from 'node:crypto';
import type { Response } from 'express';
import { GeminiApiError } from '../gemini-backend.js';
import { classifyError } from '../routes/messages.js';

/**
 * SSE レスポンスヘッダを設定する
 */
export function setupSSEHeaders(res: Response): void {
  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.flushHeaders();
}

/**
 * SSE イベントを送信する
 */
function sendSSE(res: Response, eventType: string, data: unknown): void {
  res.write(`event: ${eventType}\ndata: ${JSON.stringify(data)}\n\n`);
}

/**
 * message_start イベントを送信する
 */
function sendMessageStart(res: Response, messageId: string, model: string): void {
  sendSSE(res, 'message_start', {
    type: 'message_start',
    message: {
      id: messageId,
      type: 'message',
      role: 'assistant',
      content: [],
      model,
      stop_reason: null,
      stop_sequence: null,
      usage: {
        input_tokens: 0,
        output_tokens: 0,
      },
    },
  });
}

/**
 * content_block_start (text) イベントを送信する
 */
function sendContentBlockStart(res: Response, index: number): void {
  sendSSE(res, 'content_block_start', {
    type: 'content_block_start',
    index,
    content_block: {
      type: 'text',
      text: '',
    },
  });
}

/**
 * content_block_delta (text_delta) イベントを送信する
 */
function sendTextDelta(res: Response, index: number, text: string): void {
  sendSSE(res, 'content_block_delta', {
    type: 'content_block_delta',
    index,
    delta: {
      type: 'text_delta',
      text,
    },
  });
}

/**
 * content_block_stop イベントを送信する
 */
function sendContentBlockStop(res: Response, index: number): void {
  sendSSE(res, 'content_block_stop', {
    type: 'content_block_stop',
    index,
  });
}

/**
 * message_delta + message_stop イベントを送信してストリームを完了する
 */
function sendMessageEnd(res: Response, outputTokens: number): void {
  sendSSE(res, 'message_delta', {
    type: 'message_delta',
    delta: {
      stop_reason: 'end_turn',
      stop_sequence: null,
    },
    usage: {
      output_tokens: outputTokens,
    },
  });

  sendSSE(res, 'message_stop', {
    type: 'message_stop',
  });
}

/**
 * Child プロセスからのストリーム (ChildMessage) を Claude SSE イベントに変換して送信する
 */
export async function streamGeminiToClaudeSSE(
  childStream: AsyncGenerator<any>,
  res: Response,
  model: string,
  sessionId: string,
  sessionStore: typeof import('../session-store.js').sessionStore,
  allowedToolNames: string[]
): Promise<void> {
  const messageId = `msg_${randomUUID().replace(/-/g, '').slice(0, 24)}`;
  let blockIndex = 0;

  sendMessageStart(res, messageId, model);
  sendSSE(res, 'ping', { type: 'ping' });

  let textBlockStarted = false;
  let hasProducedAnyBlock = false;

  try {
    for await (const msg of childStream) {
      if (msg.type === 'stream_event') {
        const chunk = msg.event;
        if (chunk.type === 'content' && chunk.value) {
          if (!textBlockStarted) {
            sendContentBlockStart(res, blockIndex);
            textBlockStarted = true;
            hasProducedAnyBlock = true;
          }
          sendTextDelta(res, blockIndex, chunk.value);
        }
      } else if (msg.type === 'tool_call') {
        const callId = msg.callId;
        const name = msg.name;

        if (!allowedToolNames.includes(name)) {
          continue;
        }

        // Parent 側のセッションストアに toolCallId -> sessionId のマッピングを登録
        sessionStore.addPendingToolCall(sessionId, callId);

        if (textBlockStarted) {
          sendContentBlockStop(res, blockIndex);
          blockIndex++;
          textBlockStarted = false;
        }

        sendSSE(res, 'content_block_start', {
          type: 'content_block_start',
          index: blockIndex,
          content_block: {
            type: 'tool_use',
            id: callId,
            name: name,
            input: {},
          },
        });

        sendSSE(res, 'content_block_delta', {
          type: 'content_block_delta',
          index: blockIndex,
          delta: {
            type: 'input_json_delta',
            partial_json: JSON.stringify(msg.args),
          },
        });

        sendContentBlockStop(res, blockIndex);
        blockIndex++;
        hasProducedAnyBlock = true;
      } else if (msg.type === 'turn_end') {
        if (textBlockStarted) {
          sendContentBlockStop(res, blockIndex);
          blockIndex++;
          textBlockStarted = false;
        }

        sendSSE(res, 'message_delta', {
          type: 'message_delta',
          delta: {
            stop_reason: msg.stopReason,
            stop_sequence: null,
          },
          usage: {
            output_tokens: msg.usage?.output_tokens || 0,
          },
        });

        break; // 完全終了
      } else if (msg.type === 'error' || msg.type === 'fatal_error') {
        if (textBlockStarted) {
          sendContentBlockStop(res, blockIndex);
          textBlockStarted = false;
        }
        throw new GeminiApiError(msg.message, msg.status);
      }
    }

    if (!hasProducedAnyBlock) {
      throw new GeminiApiError('Gemini API returned an empty response', 500);
    }

  } catch (error) {
    const errorMsg = error instanceof Error ? error.message : String(error);
    console.error('[Stream Error]', errorMsg);

    const { errorType, clientMessage } = classifyError(error);

    if (!res.writableEnded) {
      const errorPayload = JSON.stringify({
        type: 'error',
        error: {
          type: errorType,
          message: clientMessage,
        },
      });
      res.write(`event: error\ndata: ${errorPayload}\n\n`);
    }
  } finally {
    if (!res.writableEnded) {
      sendSSE(res, 'message_stop', { type: 'message_stop' });
      res.end();
    }
  }
}

File: src/claude2gemini/converters/stream.py
"""Antigravity response chunks → Claude SSE events conversion."""

import json
import logging
from typing import Any, AsyncGenerator, Optional
from uuid import uuid4

logger = logging.getLogger(__name__)


async def stream_antigravity_to_claude_sse(
    response: Any,
    model: str,
    session_id: str,
    allowed_tool_names: list[str],
    pending_citations: Optional[list[dict]] = None,
) -> AsyncGenerator[str, None]:
    """Convert Antigravity ChatResponse chunks to Claude SSE event strings."""
    message_id = f"msg_{uuid4().hex[:24]}"
    block_index = 0
    msg_started = False
    estimated_tokens = 0
    text_active = False
    has_content = False
    searches = 0
    citations: list[dict] = pending_citations or []

    def msg_ensure():
        nonlocal msg_started
        if not msg_started:
            msg_started = True
            return _sse("ping", {"type": "ping"}) + _message_start(message_id, model, estimated_tokens)
        return ""

    def close_text():
        nonlocal text_active, block_index
        if text_active:
            text_active = False
            block_index += 1
            return _content_block_stop(block_index - 1)
        return ""

    try:
        async for chunk in response.chunks:
            t = chunk.get("type", "")
            val = chunk.get("value") or chunk.get("text") or ""

            if t in ("text", "content") and val:
                out = msg_ensure()
                if not text_active:
                    out += _text_start(block_index, citations)
                    text_active = True
                    has_content = True
                    citations = []
                out += _sse("content_block_delta", {
                    "type": "content_block_delta", "index": block_index,
                    "delta": {"type": "text_delta", "text": val},
                })
                yield out

            elif t == "tool_call":
                cid = chunk.get("callId") or chunk.get("id", f"toolu_{uuid4().hex[:24]}")
                nm = chunk.get("name", "unknown")
                if nm not in allowed_tool_names:
                    continue
                out = msg_ensure() + close_text()
                out += _tool_use_events(block_index, cid, nm, chunk.get("args", {}))
                block_index += 1
                has_content = True
                yield out

            elif t == "server_tool_call":
                out = msg_ensure() + close_text()
                searches += 1
                cid = chunk.get("callId") or chunk.get("id", f"srvtoolu_{uuid4().hex[:24]}")
                nm = chunk.get("name", "web_search")
                out += _server_tool_use_events(block_index, cid, nm, chunk.get("args", {}))
                block_index += 1
                has_content = True
                yield out

            elif t == "server_tool_result":
                out = msg_ensure() + close_text()
                cid = chunk.get("callId") or chunk.get("id", "")
                result = chunk.get("result", chunk.get("value", []))
                out += _sse("content_block_start", {
                    "type": "content_block_start", "index": block_index,
                    "content_block": {"type": "web_search_tool_result", "tool_use_id": cid, "content": result},
                })
                out += _content_block_stop(block_index)
                block_index += 1
                has_content = True
                if isinstance(result, list):
                    citations.extend(result)
                yield out

            elif t == "turn_end":
                if not has_content:
                    yield _error("empty_response", "Gemini API returned an empty response")
                    yield _sse("message_stop", {"type": "message_stop"})
                    return
                out = close_text()
                usage = chunk.get("usage") or chunk.get("value", {})
                du = {
                    "input_tokens": (isinstance(usage, dict) and usage.get("input_tokens", estimated_tokens)) or estimated_tokens,
                    "output_tokens": (isinstance(usage, dict) and usage.get("output_tokens", 0)) or 0,
                    "cache_read_input_tokens": (isinstance(usage, dict) and usage.get("cache_read_input_tokens", 0)) or 0,
                    "cache_creation_input_tokens": 0,
                }
                if searches > 0:
                    du["server_tool_use"] = {"web_search_requests": searches}
                out += _sse("message_delta", {
                    "type": "message_delta",
                    "delta": {"stop_reason": chunk.get("stop_reason", "end_turn"), "stop_sequence": None},
                    "usage": du,
                })
                out += _sse("message_stop", {"type": "message_stop"})
                yield out
                return

            elif t == "error":
                yield msg_ensure() + close_text()
                raise _make_stream_error(chunk)

        if not has_content:
            yield _error("empty_response", "Gemini API returned an empty response")
            yield _sse("message_stop", {"type": "message_stop"})

    except Exception as exc:
        logger.error("Stream error: %s", exc)
        ep = json.dumps({"type": "error", "error": {"type": "api_error", "message": f"Internal server error: {exc}"}})
        yield f"event: error\ndata: {ep}\n\n"
        yield _sse("message_stop", {"type": "message_stop"})


def _text_start(index: int, citations: list[dict]) -> str:
    block: dict = {"type": "text", "text": ""}
    if citations:
        block["citations"] = [
            {"type": "web_search_result_location", "url": s.get("url", ""), "title": s.get("title", ""),
             "encrypted_index": s.get("encrypted_content", ""), "cited_text": ""}
            for s in citations
        ]
    return _sse("content_block_start", {"type": "content_block_start", "index": index, "content_block": block})


def _tool_use_events(index: int, call_id: str, name: str, args: dict) -> str:
    return (
        _sse("content_block_start", {
            "type": "content_block_start", "index": index,
            "content_block": {"type": "tool_use", "id": call_id, "name": name, "input": {}},
        })
        + _sse("content_block_delta", {
            "type": "content_block_delta", "index": index,
            "delta": {"type": "input_json_delta", "partial_json": json.dumps(args)},
        })
        + _content_block_stop(index)
    )


def _server_tool_use_events(index: int, call_id: str, name: str, args: dict) -> str:
    return (
        _sse("content_block_start", {
            "type": "content_block_start", "index": index,
            "content_block": {"type": "server_tool_use", "id": call_id, "name": name, "input": {}},
        })
        + _sse("content_block_delta", {
            "type": "content_block_delta", "index": index,
            "delta": {"type": "input_json_delta", "partial_json": json.dumps(args)},
        })
        + _content_block_stop(index)
    )


def _sse(event: str, data: dict) -> str:
    return f"event: {event}\ndata: {json.dumps(data)}\n\n"


def _message_start(mid: str, model: str, tokens: int) -> str:
    return _sse("message_start", {
        "type": "message_start",
        "message": {
            "id": mid, "type": "message", "role": "assistant", "content": [], "model": model,
            "stop_reason": None, "stop_sequence": None,
            "usage": {"input_tokens": tokens, "output_tokens": 0, "cache_read_input_tokens": 0, "cache_creation_input_tokens": 0},
        },
    })


def _content_block_stop(index: int) -> str:
    return _sse("content_block_stop", {"type": "content_block_stop", "index": index})


def _error(t: str, msg: str) -> str:
    return _sse("error", {"type": "error", "error": {"type": t, "message": msg}})


class _StreamError(Exception):
    def __init__(self, message: str, status: Optional[int] = None):
        super().__init__(message)
        self.status = status


def _make_stream_error(chunk: dict) -> _StreamError:
    return _StreamError(chunk.get("message", "Unknown stream error"), chunk.get("status"))
